### DISCUSSION TITLE
Is there a way via API to get plugins and load into Mirth setup?

### DISCUSSION THREAD

**Discussion by LeviLong01** [nextgenhealthcare/connect#4723] [Q&A] [answered]
In our Docker setup, I'm trying to figure out how to programmatically pull plugins from a repo to then copy into the Docker container. 

Does NextGen provide a way to pull these plugins over an API?

A solution we are exploring is to store the plugins in S3 and use [S3 storage driver](https://docs.docker.com/registry/storage-drivers/s3/) to get these plugins.

**Comment by cturczynskyj** [comment, marked as answer]
Currently, there is no way to programmatically pull the extensions from NextGen, the Success Community doesn't support that kind of workflow. You have to grab them manually from the NextGen Success Community. We did recently add support in the official Connect Docker image for pulling extensions in as part of starting the container. Search for `EXTENSIONS_DOWNLOAD` in the [DockerHub page](https://hub.docker.com/r/nextgenhealthcare/connect). Basically, you zip up the extension zips that you grab from the Success Community, place the zip of zips somewhere that your Docker instance has access to (S3?), and set the `EXTENSIONS_DOWNLOAD` environment variable to that location and the zip will be pulled down and the extensions will be unzipped into the Connect instance in the container.

**Reply by mkopinsky** [reply]
Is `EXTENSIONS_DOWNLOAD` smart enough to download and install only the plugins corresponding to the Mirth server version? Last time I upgraded Mirth I had some downtime because I forgot to upgrade extensions and apparently Mirth 3.11 doesn't like extensions for 3.9.

**Reply by cturczynskyj** [reply]
> Is `EXTENSIONS_DOWNLOAD` smart enough to download and install only the plugins corresponding to the Mirth server version?

It does not do anything smart like that. It simply pulls down the zip that you provide, unzips it and the zips inside it into the extensions directory.

**Reply by LeviLong01** [reply]
@mkopinsky the way I'm reading @cturczynskyj's reply is that you would have to put the due diligence into making sure the right versions are in the zip file. So it's still a manual process of downloading from the website and packaging into one fat .zip file. This is still nice though so we can store .zip files somewhere and let the `EXTENSIONS_DOWNLOAD` feature do the unpacking of the .zip.

**Reply by mkopinsky** [reply]
Oh indeed. It does seem like it's a single zip file. So I would do something like:
```yaml
  mirth-connect:
    image: nextgenhealthcare/connect:3.11
    environment:
      - EXTENSIONS_DOWNLOAD=https://myserver/mirth-extensions-3.11.zip
```
Which isn't terrible - I just need to make sure to bump the image tag and the extensions zip at the same time. Still some manual work involved, but fairly foolproof if done like this.
